// cordell/lib/watcher.js
// Generated by CoffeeScript 1.4.0
(function() {
  var EventEmitter, Watcher, basename, exports, join, readdir, sep, stat, unwatchFile, watchFile, _ref, _ref1,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  EventEmitter = require('events').EventEmitter;

  _ref = require('fs'), stat = _ref.stat, readdir = _ref.readdir, watchFile = _ref.watchFile, unwatchFile = _ref.unwatchFile;

  _ref1 = require('path'), basename = _ref1.basename, join = _ref1.join;

  sep = process.platform === 'win32' ? '\\' : '/';

  Watcher = (function(_super) {

    __extends(Watcher, _super);

    function Watcher(options) {
      var _ref2, _ref3, _ref4, _ref5, _ref6, _ref7;
      if (options == null) {
        options = {};
      }
      if (options.debug != null) {
        this._debug(options.debug);
      }
      this._ignore = (_ref2 = options.ignore) != null ? _ref2 : /a^/;
      this._ignorePath = (_ref3 = options.ignorePath) != null ? _ref3 : /a^/;
      this._match = (_ref4 = options.match) != null ? _ref4 : /.*/;
      this._matchPath = (_ref5 = options.matchPath) != null ? _ref5 : /.*/;
      this._watching = {
        files: {},
        dirs: {}
      };
      this._options = {};
      this._options.persistent = (_ref6 = options.persistent) != null ? _ref6 : true;
      this._options.interval = (_ref7 = options.interval) != null ? _ref7 : 100;
    }

    Watcher.prototype._debug = function(name) {
      var debug;
      debug = require('debug')("" + name + ":watcher");
      this.on('add:file', function(path) {
        return debug("Added File: " + path);
      });
      this.on('add:dir', function(path) {
        return debug("Added Directory: " + path);
      });
      this.on('rem:file', function(path) {
        return debug("Removed File: " + path);
      });
      this.on('rem:dir', function(path) {
        return debug("Removed Directory: " + path);
      });
      this.on('change:file', function(path) {
        return debug("Changed File: " + path);
      });
      this.on('change:dir', function(path) {
        return debug("Changed Directory: " + path);
      });
      this.on('watch:file', function(path) {
        return debug("Watching File: " + path);
      });
      this.on('watch:dir', function(path) {
        return debug("Watching Directory: " + path);
      });
      this.on('unwatch:file', function(path) {
        return debug("Unwatching File: " + path);
      });
      this.on('unwatch:dir', function(path) {
        return debug("Unwatching Directory: " + path);
      });
      return this.on('error', function(path, error) {
        return debug("Error: " + path + " - " + error.message);
      });
    };

    Watcher.prototype._dir = function(path, stats) {
      var _this = this;
      if (!this._ignorePath.test(path)) {
        if (this._matchPath.test(path)) {
          return readdir(path, function(err, list) {
            if (err) {
              return _this.emit('error', path, err);
            }
            _this.emit('add', path, stats);
            _this.emit('add:dir', path, stats, list.slice(0));
            return _this.addDir(path, list);
          });
        }
      }
    };

    Watcher.prototype._file = function(path, stats) {
      if (!this._ignore.test(basename(path))) {
        if (this._match.test(basename(path))) {
          this.emit('add', path, stats);
          this.emit('add:file', path, stats);
          return this.addFile(path);
        }
      }
    };

    Watcher.prototype.add = function(path) {
      var _this = this;
      if (this._watching.dirs[path]) {
        return;
      }
      if (this._watching.files[path]) {
        return;
      }
      stat(path, function(err, stats) {
        if (err) {
          return _this.emit('error', path, err);
        }
        if (stats.isDirectory()) {
          return _this._dir(path, stats);
        }
        if (stats.isFile()) {
          return _this._file(path, stats);
        }
      });
      return this;
    };

    Watcher.prototype.addDir = function(path, list) {
      var _this = this;
      if (list == null) {
        list = [];
      }
      if (this._watching.dirs[path]) {
        return;
      }
      this._watching.dirs[path] = function(curr, prev) {
        if (curr.mtime.getTime() <= prev.mtime.getTime()) {
          return;
        }
        return readdir(path, function(err, nl) {
          var f, _i, _j, _len, _len1;
          if (err) {
            return _this.emit('error', path, err);
          }
          for (_i = 0, _len = list.length; _i < _len; _i++) {
            f = list[_i];
            if ((nl.indexOf(f)) === -1) {
              _this.rem(join(path, f));
            }
          }
          for (_j = 0, _len1 = nl.length; _j < _len1; _j++) {
            f = nl[_j];
            if ((list.indexOf(f)) === -1) {
              _this.add(join(path, f));
            }
          }
          list = nl;
          _this.emit('change', path, curr);
          return _this.emit('change:dir', path, curr, list.slice(0));
        });
      };
      watchFile(path, this._options, this._watching.dirs[path]);
      this.emit('watch', path);
      this.emit('watch:dir', path);
      return this;
    };

    Watcher.prototype.addFile = function(path) {
      var _this = this;
      if (this._watching.files[path]) {
        return;
      }
      this._watching.files[path] = function(curr, prev) {
        if (curr.mtime.getTime() <= prev.mtime.getTime()) {
          return;
        }
        _this.emit('change', path, curr);
        return _this.emit('change:file', path, curr);
      };
      watchFile(path, this._options, this._watching.files[path]);
      this.emit('watch', path);
      this.emit('watch:file', path);
      return this;
    };

    Watcher.prototype.rem = function(path) {
      var dir, file, listener, _ref2, _ref3;
      _ref2 = this._watching.files;
      for (file in _ref2) {
        if (!__hasProp.call(_ref2, file)) continue;
        listener = _ref2[file];
        if (file === path) {
          this.remFile(file);
          this.emit('rem', file);
          this.emit('rem:file', file);
          return this;
        }
        if ((file.indexOf(path + sep)) === 0) {
          this.remFile(file);
          this.emit('rem', file);
          this.emit('rem:file', file);
        }
      }
      _ref3 = this._watching.dirs;
      for (dir in _ref3) {
        if (!__hasProp.call(_ref3, dir)) continue;
        listener = _ref3[dir];
        if (dir === path || (dir.indexOf(path + sep)) === 0) {
          this.remDir(dir);
          this.emit('rem', dir);
          this.emit('rem:dir', dir);
        }
      }
      return this;
    };

    Watcher.prototype.remDir = function(path) {
      if (!this._watching.dirs[path]) {
        return;
      }
      unwatchFile(path, this._watching.dirs[path]);
      delete this._watching.dirs[path];
      this.emit('unwatch', path);
      this.emit('unwatch:dir', path);
      return this;
    };

    Watcher.prototype.remFile = function(path) {
      if (!this._watching.files[path]) {
        return;
      }
      unwatchFile(path, this._watching.files[path]);
      delete this._watching.files[path];
      this.emit('unwatch', path);
      this.emit('unwatch:file', path);
      return this;
    };

    Watcher.prototype.close = function() {
      var dirs, files, listener, path;
      files = this._watching.files;
      dirs = this._watching.dirs;
      this._watching.files = {};
      this._watching.dirs = {};
      for (path in files) {
        if (!__hasProp.call(files, path)) continue;
        listener = files[path];
        unwatchFile(path, listener);
        this.emit('unwatch', path);
        this.emit('unwatch:file', path);
      }
      for (path in dirs) {
        if (!__hasProp.call(dirs, path)) continue;
        listener = dirs[path];
        unwatchFile(path, listener);
        this.emit('unwatch', path);
        this.emit('unwatch:dir', path);
      }
      return this;
    };

    return Watcher;

  })(EventEmitter);

  module.exports = exports = Watcher;

}).call(this);
